sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SchemaPruning.scala

```diff
@@ -103,8 +104,14 @@ object SchemaPruning extends Rule[LogicalPlan] {
     }
   }
 
+  private def containsNestedColumn(fsRelation: HadoopFsRelation): Boolean =
```

Comment:
> does nested column pruning work for metadata columns? 

Yes, we combined them in this logic. And the metadata pruning should work for every file format: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SchemaPruning.scala#L92
(It's hidden file metadata for file-based V1 source)

Reply:
Are you sure?

I run `FileMetadataStructSuite` test with different pruning configurations:

`SQLConf.NESTED_SCHEMA_PRUNING_ENABLED` = `false`:

```
== Physical Plan ==
*(1) Project [name#1884, age#1885, info#1886.id AS id#1890L, _metadata#1891.file_name AS file_name#1892]
+- *(1) Project [name#1884, age#1885, info#1886, named_struct(file_name, file_name#1902) AS _metadata#1891]
   +- *(1) ColumnarToRow
      +- FileScan parquet [name#1884,age#1885,info#1886,file_name#1902] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex(2 paths)[file:/spark/target/tmp/spark-9e425a7e-5b6c-46aa..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<name:string,age:int,info:struct<id:bigint,university:string>>

```

`SQLConf.NESTED_SCHEMA_PRUNING_ENABLED` = `true`:

```
== Physical Plan ==
*(1) Project [name#1884, age#1885, info#1886.id AS id#1890L, _metadata#1891.file_name AS file_name#1892]
+- *(1) Project [name#1884, age#1885, info#1886, named_struct(file_name, file_name#1902) AS _metadata#1891]
   +- *(1) ColumnarToRow
      +- FileScan parquet [name#1884,age#1885,info#1886,file_name#1902] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex(2 paths)[file:/spark/target/tmp/spark-a0afa17b-52e7-4d1c..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<name:string,age:int,info:struct<id:bigint>>
```

You can see `info` is pruned. But for `_metadata` column, it is always pruned and no difference between pruning/non-pruning cases.

Reply:
`_metadata` is added by datasource based on if it is required in the query. It is not a physical column that we prune from Parquet by changing the required schema. I don't think it follows the same logic there.

Reply:
We combined them in the same logic at `prunePhysicalColumns(...)`. And the flag `NESTED_SCHEMA_PRUNING_ENABLED` controls only data schema ([here](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SchemaPruning.scala#L73)), but not `_metadata`. 

Basically, we try to prune `_metadata` regardless of the flag and the file format.

Reply:
Oh, I see. Operators will collect metadata column names. And during query execution, it will update into a row based on these column names. In `prunePhysicalColumns`, no matter the config value is, it always prune metadata columns.

Reply:
Yes, exactly!

Reply:
Thank you all.